TRT: remove serialized native function from TRT Engine Op

phillip-kravtsov opened the pull request:
Removing the serialized fallback from the TRTEngineOp, and removed the option to not have a native fallback. This gets around a protobuf size limit for large models (BERT-Large).

Commit:
Inital commit: removed serialized string from dynamic TRT engine.

Commit:
Inital commit: removed serialized string from dynamic TRT engine.

Commit:
Inital commit: removed serialized string from dynamic TRT engine.

Added error checking in trt_engine_op.cc

Removed use_function_backup parameter.

Removed excessively verbose logging from trt.

Mild cleanup.

More mild cleanup, removed unnecessary static condition.y

Moved constant IO strings into class. Renamed method in funcdef_to_graphdef. Formatted, removed commenting.

Removed duplicate function in trt_engine_op.cc

Inital commit: removed serialized string from dynamic TRT engine.

Removed use_function_backup parameter.

Removed excessively verbose logging from trt.

Mild cleanup.

More mild cleanup, removed unnecessary static condition.y

Moved constant IO strings into class. Renamed method in funcdef_to_graphdef. Formatted, removed commenting.

Removed duplicate function in trt_engine_op.cc

Removed commented out code.

Reverted unnecessary formatting adjustments.

Removed rest of unnecessary formatting.

Renamed prefixes and sgraph --> segment_graph.

TRT minor improvements & correctionx

TFTRT: Changed segment to graphdef conversion to create arg/ret ops instead of placeholder/identity.

TF-TRT: ySimplified graph conversion functions

Fixed rebase-induced issues

Commit:
Added error checking in trt_engine_op.cc

Commit:
Added error checking in trt_engine_op.cc

Commit:
Removed use_function_backup parameter.

Commit:
Removed use_function_backup parameter.

Commit:
Removed excessively verbose logging from trt.

Commit:
Removed excessively verbose logging from trt.

Commit:
Mild cleanup.

Commit:
Mild cleanup.

Commit:
More mild cleanup, removed unnecessary static condition.y

Commit:
More mild cleanup, removed unnecessary static condition.y

Commit:
Moved constant IO strings into class. Renamed method in funcdef_to_graphdef. Formatted, removed commenting.

Commit:
Moved constant IO strings into class. Renamed method in funcdef_to_graphdef. Formatted, removed commenting.

Commit:
Removed duplicate function in trt_engine_op.cc

Commit:
Removed duplicate function in trt_engine_op.cc

Commit:
Merge branch 'trt_remove_serialized_from_op' of https://github.com/phillip-kravtsov/tensorflow into trt_remove_serialized_from_op

gbaned commented:
@phillip-kravtsov Please sign CLA in order to proceed with next steps. Thanks!

Commit:
Removed commented out code.

aaroey commented:
@phillip-kravtsov please sign the cla.

phillip-kravtsov commented:
Signed the NVIDIA Corporate CLA!

aaroey commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -363,9 +361,9 @@ string DebugString(const nvinfer1::Permutation& permutation, int len) {
-                ", name=", tensor.getName(),
-                ", dtype=", DebugString(tensor.getType()),
-                ", dims=", DebugString(tensor.getDimensions()), ")");
+                ", name=", tensor.getName(), ", dtype=",
```
Can you revert changes like this in this PR? They contributed to a big part of the diff in this and other files.

aaroey commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -1460,7 +1455,7 @@ void Converter::MaybeApplyQuantizationRanges() {
   // Infer ranges across marked ops.
   PropagateQuantizationRanges();
-  // Apply ranges.
+// Apply ranges.
```
Please revert.

aaroey commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -1515,19 +1510,15 @@ void Converter::MaybeApplyQuantizationRanges() {
-           IsConvolution,
-           IsScale,
-           IsClipOrRelu,
+           IsConvolution, IsScale, IsClipOrRelu,
```
Please revert.

aaroey commented on tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -179,9 +192,10 @@ void* GetTensorAddress(const Tensor* tensor_ptr) {
+                                            const string& device_name) {
   VLOG(1) << "Constructing function handle";
-  auto lib = ctx->function_library();
+  // auto lib = ctx->function_library();
```
Remove

aaroey commented on tensorflow/compiler/tf2tensorrt/utils/funcdef_to_graphdef.h:
```diff
@@ -0,0 +1,45 @@
+
+string AppendIdToNodeName(const Node* n);
+
+void ToGraphDefWithIOPrefix(const Graph* g, GraphDef* gdef);
```
Please add some unit tests for these functions.

aaroey reviewed (changes requested):
Just some high level comments, please help to fix them and I'll review the details later.

Commit:
Reverted unnecessary formatting adjustments.

Commit:
Removed rest of unnecessary formatting.

phillip-kravtsov commented:
@aaroey I've reverted the extraneous formatting changes and made the changes you requested, except for the unit tests. The functionality of the `ToGraphDefWithIOPrefix` is the same as the functionality as in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/function.cc#L2192 , but with a different name remapping. 
Would it make more sense to change the one in function.cc to accept any renaming function, or keep the duplicate?

aaroey commented on tensorflow/compiler/tf2tensorrt/utils/funcdef_to_graphdef.cc:
```diff
@@ -0,0 +1,142 @@
+    auto id = n->id();
+    if (n->IsArg()) {
+      VLOG(2) << "Arg Node id used for unique naming is " << id;
+      input_node_ids->push_back(id);
```
`graph` is a temporary object, why do we use the node id as the id of the input/output nodes?

aaroey commented on tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -309,7 +325,9 @@ void TRTEngineOp::ExecuteCalibration(OpKernelContext* ctx,
     CHECK_EQ(t.TotalBytes(), device_tensor->TotalBytes());
-    input_data.emplace(StrCat(kInputPHName, i), data_address);
+    input_data.emplace(
+        StrCat(prefixes.kInputPHName, static_engine_ ? i : input_node_ids_[i]),
```
I don't understand, why not just use the same id `i`? ConvertGraphDefToEngine() uses the same `i` from the Placeholder/Identity node's name suffix right?

Same question below.

aaroey reviewed (changes requested):
Thanks for the fixes. Regarding your question about `ToGraphDefWithIOPrefix`, I don't understand why we need a different renaming, is it because the function library converts all name to lowercase? If so can we reuse `ToGraphDef()` and do the renaming afterwards?

Commit:
Renamed prefixes and sgraph --> segment_graph.

Commit:
TRT minor improvements & correctionx

gbaned commented:
@phillip-kravtsov Could you please resolve the conflicts? Thanks!

Commit:
TFTRT: Changed segment to graphdef conversion to create arg/ret ops instead of placeholder/identity.

Commit:
TF-TRT: ySimplified graph conversion functions

phillip-kravtsov commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -5016,19 +5014,34 @@ Status ConvertGraphDefToEngine(
-        return errors::InvalidArgument("Failed to parse slot number from ",
-                                       node_name);
+      string type_key;
+      if (node_def.op() == "Placeholder") {
```
In the static case, CreateTRTNode uses the original segment graphdef, not the one recreated from the function, so I believe the graph at that point has those nodes, not _Arg/_Retval.

phillip-kravtsov commented on tensorflow/compiler/tf2tensorrt/utils/funcdef_to_graphdef.cc:
```diff
@@ -0,0 +1,142 @@
+
+auto prefixes = IONamePrefixes();
+
+string AppendIdToNodeName(const Node* n) {
```
The Graph-->GraphDef conversion in funcdef_to_graphdef.cc mirrors the behavior of the conversion in core/common_runtime/function.cc, I retained the uniquification of nodes (which uses the id's to rename them). I'm unsure if this is necessary or not, if we can be sure that the node names are already unique, we can remove the remapping (otherwise, moving it outside of the conversion). 
The special remapping is needed because the code identifies I/O nodes by prefix, while the method in core/common_runtime/function.cc simply renames all nodes by their id, removing prefixes.

phillip-kravtsov commented:
@aaroey Thank you for the thorough review! I've made the major changes requested, excepting a couple of the concerns, on which I've commented above. I've also rebased the commits into one, and rebased onto master.

aaroey committed:
Merge with upstream

aaroey committed:
Merge branch 'trt_remove_serialized_from_op' of https://github.com/phillip-kravtsov/tensorflow into trt_remove_serialized_from_op

aaroey committed:
Multiple changes:
- Add a new segment_func attribute for the segment function, which is required to get all v2 tests passed
- Remove ConvertSegmentToGraph()
- Inline RegisterSegmentToFunctionLibrary()
- Use connection.port_number to set the index of _Arg/_Retval node
- Update FunctionDefToGraphDef() to use graph->ToGraphDef(), so we don't need ToGraphDefWithIOPrefix()
- Remove funcdef_to_graphdef.{h,cc} and inline the only function FunctionDefToGraphDef() to trt_engine_op.cc
- Fix corresponding tests
- Stop setting input/output shapes attributes which are deprecated
- Fix formatting

aaroey reviewed (changes requested):
Thanks for the fixes. In future PRs please don't do rebase/force push, it makes it very hard or impossible to see the diff since last review.

aaroey committed:
Remove funcdef_to_graphdef.{h,cc}
